src/main/java/org/wildfly/security/auth/provider/oauth2/OAuth2SecurityRealm.java

```diff
@@ -0,0 +1,313 @@
+/*
+ * JBoss, Home of Professional Open Source.
+ * Copyright 2016 Red Hat, Inc., and individual contributors
+ * as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.wildfly.security.auth.provider.oauth2;
+
+import org.wildfly.common.Assert;
+import org.wildfly.security._private.ElytronMessages;
+import org.wildfly.security.auth.principal.NamePrincipal;
+import org.wildfly.security.auth.server.RealmIdentity;
+import org.wildfly.security.auth.server.RealmUnavailableException;
+import org.wildfly.security.auth.server.SecurityRealm;
+import org.wildfly.security.auth.server.SupportLevel;
+import org.wildfly.security.authz.Attributes;
+import org.wildfly.security.authz.AuthorizationIdentity;
+import org.wildfly.security.credential.Credential;
+import org.wildfly.security.evidence.BearerTokenEvidence;
+import org.wildfly.security.evidence.Evidence;
+
+import javax.json.JsonObject;
+import javax.net.ssl.HostnameVerifier;
+import javax.net.ssl.SSLContext;
+import java.net.URL;
+import java.security.Principal;
+
+import static org.wildfly.security.auth.provider.oauth2.OAuth2Util.introspectAccessToken;
+import static org.wildfly.security.auth.provider.oauth2.OAuth2Util.toAttributes;
+
+/**
+ * An oAuth2-backed {@link SecurityRealm} based on RFC-7662 in order to verify bearer tokens from a given {@link BearerTokenEvidence}
+ * against a token introspection endpoint and build identities based on the token metadata or claims.
+ *
+ * @author <a href="mailto:psilva@redhat.com">Pedro Igor</a>
+ */
+public class OAuth2SecurityRealm implements SecurityRealm {
+
+    private final Builder.OAuth2Configuration configuration;
+
+    public static Builder builder() {
+        return new Builder();
+    }
+
+    public OAuth2SecurityRealm(Builder.OAuth2Configuration configuration) {
```

Comment:
This constructor should not be ``public`` given that ``OAuth2Configuration`` is not.